rust/lance-index/src/scalar/inverted/scorer.rs

```diff
@@ -22,13 +22,106 @@ pub trait Scorer: Send + Sync {
 pub const K1: f32 = 1.2;
 pub const B: f32 = 0.75;
 
-pub struct BM25Scorer<'a> {
+pub enum BM25Scorer<'a> {
+    Index(IndexBM25Scorer<'a>),
+    Memory(MemoryBM25Scorer),
+    Union(Box<BM25Scorer<'a>>, Box<BM25Scorer<'a>>),
+}
+
+impl<'a> BM25Scorer<'a> {
+    pub fn num_docs(&self) -> usize {
+        match self {
+            BM25Scorer::Index(scorer) => scorer.num_docs(),
+            BM25Scorer::Memory(scorer) => scorer.num_docs(),
+            BM25Scorer::Union(left, right) => left.num_docs() + right.num_docs(),
+        }
+    }
+
+    pub fn avgdl(&self) -> f32 {
+        match self {
+            BM25Scorer::Index(scorer) => scorer.avgdl(),
+            BM25Scorer::Memory(scorer) => scorer.avgdl(),
+            BM25Scorer::Union(left, right) => {
+                (left.avgdl() * left.num_docs() as f32 + right.avgdl() * right.num_docs() as f32)
+                    / (left.num_docs() + right.num_docs()) as f32
+            }
+        }
+    }
+
+    // the number of documents that contain the token
+    pub fn nq(&self, token: &str) -> usize {
```

Comment:
Same thing here about the method name. I like that you have a comment describing it though.